### PR TITLE
Fuzz inorder undo

### PR DESCRIPTION
**Pull request by daesunp**
## Description

This PR adds targeted fuzz tests for inorder undo and redo.

### CODE REVIEW

File: experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts [hunk 1]
```diff
@@ -172,4 +173,100 @@ describe("Fuzz - Targeted", () => {
...
+			for (const [i, client] of clients.entries()){
+				const tree = client.channel;
+
+				// save final tree states to validate redo later
+				finalTreeStates.push(toJsonableTree(tree))
+
+				// call undo() until tree no more edits left
+				for (let j = 0; j<opsPerRun; j++){
+					tree.undo()
+				}
```
Comment (daesunp) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts:222]: @yann-achard-MS I am currently just using a loop and calling the `tree.undo()` method `opsPerRun` number of times. This is a more "lazy" way of doing this, as we only produce at max this many number of ops. But would there be a better way to do this? One way would be to check the `UndoRedoManager` in the tree's branch to check if the head is undefined, but we do not have access to this as it is not exposed.

**Comment by daesunp** [comment]
Currently the tests seem to be failing when synchronization is enabled. The test seems to be passing when I use a single client, or manually remove the code for `mixinSynchronization` in the `DDSFuzzHarness.ts`. Should we skip this test for now? or keep it enabled with a single client?

@Abe27342 Is there a way of passing in an argument to prevent synchronization between clients?

**Comment by Abe27342** [comment]
> Currently the tests seem to be failing when synchronization is enabled. The test seems to be passing when I use a single client, or manually remove the code for `mixinSynchronization` in the `DDSFuzzHarness.ts`. Should we skip this test for now? or keep it enabled with a single client?
> 
> @Abe27342 Is there a way of passing in an argument to prevent synchronization between clients?

no, there isn't such a way and it doesn't make much sense to add one--the harness's purpose is to test eventual consistency across clients, which you need synchronization for. There are probably legitimate eventual consistency bugs, and checking in the test skipped (or just some seeds skipped) still provides value as a means of something to work at improving.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts [hunk 2]
```diff
@@ -172,4 +173,100 @@ describe("Fuzz - Targeted", () => {
...
+				assert(finalState.initialTreeState !== undefined);
+				validateTree(client.channel, finalState.initialTreeState)
```
Comment (Abe27342) [inline experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts:224]: don't you need to synchronize client state before all trees will have arrived back at their original value?

File: experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts (hunk 1 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts:222]: SharedTree is supposed to be exposing an undo stack size ([work item](https://dev.azure.com/fluidframework/internal/_workitems/edit/3625)) so I'm reticent to invest too much in something more refined than what you have here.

**Comment by yann-achard-MS** [comment]
> > Currently the tests seem to be failing when synchronization is enabled. The test seems to be passing when I use a single client, or manually remove the code for `mixinSynchronization` in the `DDSFuzzHarness.ts`. Should we skip this test for now? or keep it enabled with a single client?
> 
> There are probably legitimate eventual consistency bugs, and checking in the test skipped (or just some seeds skipped) still provides value as a means of something to work at improving.

I think it's fine to commit the test as skipped for now.

File: experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts (hunk 2 above)
Comment (daesunp) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts]: synchronized after calling `undo()` on all clients

File: experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts (hunk 1 above)
Comment (Abe27342) [reply experimental/dds/tree2/src/test/shared-tree/fuzz/targetedFuzz.spec.ts:222]: seems fine to me, assuming it's part of the spec that undo no-ops when it reaches an empty stack.

**Review by Abe27342 — APPROVED** [review]

**Pull request merged**